hazelcast/src/test/java/com/hazelcast/instance/NodeThreadLeakTest.java

```diff
@@ -44,43 +44,51 @@
 /**
  * Tests that an exception in the {@link Node} and {@link NodeEngineImpl} constructor leads to properly finished services.
  */
-@RunWith(PowerMockRunner.class)
-@PrepareForTest({DefaultNodeContext.class, NodeEngineImpl.class})
-@Category(QuickTest.class)
+@RunWith(HazelcastParallelClassRunner.class)
+@Category({QuickTest.class, ParallelJVMTest.class})
 public class NodeThreadLeakTest extends HazelcastTestSupport {
 
     private static final HazelcastException HAZELCAST_EXCEPTION
             = new HazelcastException("Test exception - Emulates service failure.");
 
     @Test
     public void testLeakWhenCreatingConnectionManager() throws Exception {
-        mockConstructorAndTest(TcpServer.class);
+        testFailingHazelcastCreation(new DefaultNodeContext() {
+            @Override
+            public Server createServer(Node node, ServerSocketRegistry registry, LocalAddressRegistry addressRegistry) {
+                super.createServer(node, registry, addressRegistry);
+                throw HAZELCAST_EXCEPTION;
+            }
+        });
     }
 
     @Test
-    public void testCreatingEventServiceImplFails() throws Exception {
-        mockConstructorAndTest(EventServiceImpl.class);
+    public void testFailingInNodeEngineImplConstructor() throws Exception {
+        testFailingHazelcastCreation(new DefaultNodeContext() {
+            @Override
+            public NodeExtension createNodeExtension(Node node) {
+                return new DefaultNodeExtension(node) {
+                    @Override
+                    public <T> T createService(Class<T> clazz, Object... params) {
+                        throw HAZELCAST_EXCEPTION;
+                    }
+                };
+            }
+        });
     }
 
-    @Test
-    public void testCreatingExecutionServiceImplFails() throws Exception {
-        mockConstructorAndTest(ExecutionServiceImpl.class);
-    }
-
-    @Test
-    public void testCreatingOperationParkerImplFails() throws Exception {
-        mockConstructorAndTest(OperationParkerImpl.class);
-    }
```

Comment:
Are you sure that we still cover the same code paths? Previously there were 4 tests, now there are 2.

Reply:
I think they were testing the same thing: what happens when these constructors fail in the NodeEngineImpl constructor. As you can see all of them are create in the same place https://github.com/hazelcast/hazelcast/blob/77ef2bfdfb19005b560cefb2a104994019983d2a/hazelcast/src/main/java/com/hazelcast/spi/impl/NodeEngineImpl.java#L145-L148

So instead of checking them separately (which would be hard without reflection hacks) I've decided to throw an exception here: https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/spi/impl/NodeEngineImpl.java#L150.

Reply:
Also added extra checks to make sure we're in the right state before throwing test exception: https://github.com/hazelcast/hazelcast/pull/20295/commits/c314c5303b21b7641062d475099c7d94749977f3